File: app/src/test/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagesearch/searchbox/SearchBoxViewModelTest.java
package com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.searchbox;

import android.app.Application;
import android.text.TextUtils;
import android.util.Log;

import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.Observer;

import com.ch.yoon.kakao.pay.imagesearch.R;
import com.ch.yoon.kakao.pay.imagesearch.RxSchedulerRule;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.stubbing.Answer;
import org.powermock.core.classloader.annotations.PrepareForTest;
import org.powermock.modules.junit4.PowerMockRunner;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import kotlin.Unit;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.powermock.api.mockito.PowerMockito.mockStatic;
import static org.powermock.api.mockito.PowerMockito.when;

@RunWith(PowerMockRunner.class)
@PrepareForTest({Log.class, TextUtils.class})
public class SearchBoxViewModelTest {

    private static final String DELETE_SUCCESS_MESSAGE = "삭제 완료";
    private static final String EMPTY_KEYWORD_MESSAGE = "검색어를 입력해주세요";

    @Rule
    public InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();
    @Rule
    public RxSchedulerRule rxSchedulerRule = new RxSchedulerRule();

    @Mock
    private Application mockApplication;
    @Mock
    private ImageSearchRepository mockImageSearchRepository;
    @Mock
    private Observer<Unit> mockVoidObserver;

    private SearchBoxViewModel searchBoxViewModel;

    @Before
    public void init() {
        MockitoAnnotations.initMocks(this);

        initApplication();
        initSearchBoxViewModel();
        initUtils();
    }

    private void initApplication() {
        when(mockApplication.getString(R.string.success_all_image_info_delete_by_keyword)).thenReturn(DELETE_SUCCESS_MESSAGE);
        when(mockApplication.getString(R.string.empty_keyword_guide)).thenReturn(EMPTY_KEYWORD_MESSAGE);
    }

    private void initSearchBoxViewModel() {
        searchBoxViewModel = new SearchBoxViewModel(mockApplication, mockImageSearchRepository);
    }

    private void initUtils() {
        mockStatic(Log.class);
        when(Log.d(any(String.class), any(String.class))).thenReturn(0);
        
        mockStatic(TextUtils.class);
        when(TextUtils.isEmpty(any(CharSequence.class))).thenAnswer((Answer<Boolean>) invocation -> {
            CharSequence text = invocation.getArgument(0);
            if(text instanceof String) {
                return text.length() == 0;
            } else {
                return false;
            }
        });
    }

    @Test
    public void 키워드_검색_버튼_클릭시_입력한_키워드가_반영되는지_테스트() {
        // given
        when(mockImageSearchRepository.insertOrUpdateSearchLog(any(String.class)))
                .thenReturn(Single.just(new SearchLog("테스트", 1)));

        // when
        searchBoxViewModel.onClickSearchButton("테스트");

        // then
        assertEquals(searchBoxViewModel.getSearchKeyword().getValue(), "테스트");
    }

    @Test
    public void 키워드_검색_버튼_클릭시_키워드가_비어있다면_거절_메시지가_반영되는지_테스트() {
        // when
        searchBoxViewModel.onClickSearchButton("");

        // then
        assertEquals(searchBoxViewModel.getShowMessageEvent().getValue(), EMPTY_KEYWORD_MESSAGE);
    }

    @Test
    public void 이미지_로드가_되는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(createVirtualSearchLogList(3)));

        // when
        searchBoxViewModel.loadSearchLogList();

        // then
        searchBoxViewModel.getSearchLogList().observeForever(receivedList -> {
            assertEquals(receivedList.size(), 3);
        });
    }

    @Test
    public void 이미지_로드시_내림차순으로_정렬되는지_테스트() {
        // given
        List<SearchLog> searchLogList = createVirtualSearchLogList(3);

        List<SearchLog> expectedList = createVirtualSearchLogList(3);
        Collections.sort(expectedList);

        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(searchLogList));

        // when
        searchBoxViewModel.loadSearchLogList();

        // then
        searchBoxViewModel.getSearchLogList().observeForever(receivedList -> {
            assertEquals(receivedList, expectedList);
        });

    }

    @Test
    public void 검색상자_클릭시_포커스를_갖는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        // when
        searchBoxViewModel.onClickSearchBox();

        //then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), true);
    }

    @Test
    public void 키워드_검색_버튼_클릭시_키워드가_비어있지_않다면_포커스를_잃는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        when(mockImageSearchRepository.insertOrUpdateSearchLog(any(String.class)))
                .thenReturn(Single.just(new SearchLog("테스트", 1)));

        // when
        searchBoxViewModel.loadSearchLogList();
        searchBoxViewModel.onClickSearchBox();
        searchBoxViewModel.onClickSearchButton("테스트");

        // then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), false);
    }

    @Test
    public void 리스트가_비어있어도_포커스를_갖는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        // when
        searchBoxViewModel.loadSearchLogList();
        searchBoxViewModel.onClickSearchBox();

        // then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), true);
    }

    @Test
    public void 키워드_검색_버튼_클릭시_키워드가_비어있어도_갖고있던_포커스를_계속_유지하는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        // when
        searchBoxViewModel.loadSearchLogList();
        searchBoxViewModel.onClickSearchBox();
        searchBoxViewModel.onClickSearchButton("");

        //then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), true);
    }

    @Test
    public void 배경_클릭시_포커스를_잃는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        // when
        searchBoxViewModel.onClickSearchBox();
        searchBoxViewModel.onClickBackground();

        //then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), false);
    }

    @Test
    public void 뒤로가기_클릭시_갖고있던_포커스를_잃는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));

        // when
        searchBoxViewModel.onClickSearchBox();
        searchBoxViewModel.onClickBackPressButton();

        // then
        assertEquals(searchBoxViewModel.getSearchBoxFocus().getValue(), false);
    }

    @Test
    public void 뒤로가기_두번_클릭시_포커스가_존재했다면_종료_이벤트가_호출되는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(new ArrayList<>()));
        searchBoxViewModel.getSearchBoxFinishEvent().observeForever(mockVoidObserver);

        // when
        searchBoxViewModel.onClickSearchBox();
        searchBoxViewModel.onClickBackPressButton();
        searchBoxViewModel.onClickBackPressButton();

        // then
        verify(mockVoidObserver, times(1)).onChanged(any());
    }

    @Test
    public void 뒤로가기_클릭시_포커스가_없다면_종료_이벤트가_호출되는지_테스트() {
        // given
        searchBoxViewModel.getSearchBoxFinishEvent().observeForever(mockVoidObserver);

        // when
        searchBoxViewModel.onClickBackPressButton();

        // then
        verify(mockVoidObserver, times(1)).onChanged(any());
    }

    @Test
    public void 키워드_삭제_버튼_클릭시_레파지토리에_삭제_요청을_하는지_테스트() {
        // given
        when(mockImageSearchRepository.deleteSearchLog("테스트"))
            .thenReturn(Completable.complete());

        // when
        searchBoxViewModel.onClickSearchLogDeleteButton(new SearchLog("테스트", 1));

        // then
        verify(mockImageSearchRepository, times(1))
            .deleteSearchLog("테스트");
    }

    @Test
    public void 키워드_삭제_버튼_클릭시_보유하던_검색_기록_목록에서_제거하는지_테스트() {
        // given
        List<SearchLog> searchLogList = createVirtualSearchLogList(3);
        List<SearchLog> expectedList = createVirtualSearchLogList(3);
        expectedList.remove(0);
        Collections.sort(expectedList);

        when(mockImageSearchRepository.requestSearchLogList()).thenReturn(Single.just(searchLogList));

        when(mockImageSearchRepository.deleteSearchLog(any(String.class)))
            .thenReturn(Completable.complete());

        // when
        searchBoxViewModel.loadSearchLogList();
        searchBoxViewModel.onClickSearchLogDeleteButton(searchLogList.get(0));

        // then
        searchBoxViewModel.getSearchLogList().observeForever(receivedList -> {
            assertEquals(expectedList, receivedList);
        });
    }

    @Test
    public void 키워드_검색_버튼_클릭시_기존에_검색했던_키워드라면_목록의_가장_앞쪽으로_이동시키는지_테스트() {
        // given
        when(mockImageSearchRepository.requestSearchLogList())
            .thenReturn(Single.just(createVirtualSearchLogList(3)));

        when(mockImageSearchRepository.insertOrUpdateSearchLog("테스트0"))
            .thenReturn(Single.just(new SearchLog("테스트0", 4)));

        // when
        searchBoxViewModel.loadSearchLogList();
        searchBoxViewModel.onClickSearchButton("테스트0");

        // then
        searchBoxViewModel.getSearchLogList().observeForever(searchLogs -> {
            assertEquals(searchLogs.get(0).getKeyword(), "테스트0");
            assertEquals(searchLogs.size(), 3);
        });
    }

    private List<SearchLog> createVirtualSearchLogList(int size) {
        ArrayList<SearchLog> searchLogList = new ArrayList<>();
        for(int i=0; i<size; i++) {
            searchLogList.add(new SearchLog("테스트" + i, i));
        }

        return searchLogList;
    }

}

File: app/src/test/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagesearch/searchbox/SearchBoxViewModelTest.kt
package com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.searchbox

import android.app.Application
import android.util.Log
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import com.belongings.bag.belongingsbag.RxSchedulerRule
import com.ch.yoon.kakao.pay.imagesearch.R
import com.ch.yoon.kakao.pay.imagesearch.data.repository.ImageRepository
import com.ch.yoon.kakao.pay.imagesearch.data.repository.model.SearchLog
import io.mockk.MockKAnnotations
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.mockkStatic
import io.mockk.verify
import io.reactivex.Completable
import io.reactivex.Single
import junit.framework.Assert.assertEquals
import net.bytebuddy.matcher.ElementMatchers.returns
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import java.util.*

/**
 * Creator : ch-yoon
 * Date : 2019-11-01.
 */
class SearchBoxViewModelTest {

    companion object {
        private const val DELETE_SUCCESS_MESSAGE = "삭제 완료"
        private const val EMPTY_KEYWORD_MESSAGE = "검색어를 입력해주세요"
    }

    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()
    @get:Rule
    val rxSchedulerRule = RxSchedulerRule()

    @MockK
    lateinit var mockApplication: Application
    @MockK
    lateinit var mockImageSearchRepository: ImageRepository

    private lateinit var searchBoxViewModel: SearchBoxViewModel


    @Before
    fun init() {
        MockKAnnotations.init(this, relaxUnitFun = true)

        initApplication()
        initUtils()
        initSearchBoxViewModel()
    }

    private fun initApplication() {
        every { mockApplication.getString(R.string.success_all_image_info_delete_by_keyword) } returns DELETE_SUCCESS_MESSAGE
        every { mockApplication.getString(R.string.empty_keyword_guide) } returns EMPTY_KEYWORD_MESSAGE
    }

    private fun initUtils() {
        mockkStatic(Log::class)
        every { Log.d(any(), any()) } returns -1
    }

    private fun initSearchBoxViewModel() {
        searchBoxViewModel = SearchBoxViewModel(mockApplication, mockImageSearchRepository)
    }

    @Test
    fun `이미지 로드가 되는지 테스트`() {
        // given
        every {mockImageSearchRepository.requestSearchLogList() } returns (Single.just(createVirtualSearchLogList(3)));

        // when
        searchBoxViewModel.loadSearchLogList()

        // then
        searchBoxViewModel.searchLogList.observeForever { receivedList ->
            assertEquals(3, receivedList.size)
        }
    }

    @Test
    fun `키워드 검색 버튼 클릭시 입력한 키워드가 반영되는지 테스트`() {
        // given
        every { mockImageSearchRepository.insertOrUpdateSearchLog(any()) } returns (Single.just(SearchLog("테스트", 1)))

        // when
        searchBoxViewModel.onClickSearchButton("테스트")

        // then
        searchBoxViewModel.searchKeyword.observeForever { keyword ->
            assertEquals("테스트", keyword)
        }
    }

    @Test
    fun `키워드 검색 버튼 클릭시 키워드가 비어있다면 거절 메시지가 반영되는지 테스트`() {
        // when
        searchBoxViewModel.onClickSearchButton("")

        // then
        searchBoxViewModel.showMessageEvent.observeForever { message ->
            assertEquals(EMPTY_KEYWORD_MESSAGE, message)
        }
    }

    @Test
    fun `이미지 로드시 내림차순으로 정렬되는지 테스트`() {
        // given
        val searchLogList = createVirtualSearchLogList(3)
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(searchLogList))

        // when
        searchBoxViewModel.loadSearchLogList()

        // then
        val expectedList = createVirtualSearchLogList(3).sorted()
        searchBoxViewModel.searchLogList.observeForever { receivedList ->
            assertEquals(expectedList, receivedList)
        }
    }

    @Test
    fun `검색상자 클릭시 포커스를 갖는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        searchBoxViewModel.onClickSearchBox()

        //then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(true, focus)
        }
    }

    @Test
    fun `키워드 검색 버튼 클릭시 키워드가 비어있지 않다면 포커스를 잃는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))
        every { mockImageSearchRepository.insertOrUpdateSearchLog(any()) } returns (Single.just(SearchLog("테스트", 1)))

        // when
        searchBoxViewModel.loadSearchLogList()
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickSearchButton("테스트")

        // then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(false, focus)
        }
    }

    @Test
    fun `키워드 검색 버튼 클릭시 키워드가 비어있다면 포커스를 계속 유지하는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        searchBoxViewModel.loadSearchLogList()
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickSearchButton("")

        //then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(true, focus)
        }
    }

    @Test
    fun `리스트가 비어있어도 포커스를 갖는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        searchBoxViewModel.loadSearchLogList()
        searchBoxViewModel.onClickSearchBox()

        // then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(true, focus)
        }
    }

    @Test
    fun `배경 클릭시 갖고 있던 포커스를 잃는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickBackground()

        //then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(false, focus)
        }
    }

    @Test
    fun `뒤로가기 클릭시 갖고있던 포커스를 잃는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickBackPressButton()

        // then
        searchBoxViewModel.searchBoxFocus.observeForever { focus ->
            assertEquals(false, focus)
        }
    }

    @Test
    fun `뒤로가기 클릭시 포커스가 없다면 종료 이벤트가 호출되는지 테스트`() {
        // when
        var finishEventCount = 0
        searchBoxViewModel.searchBoxFinishEvent.observeForever { finishEventCount++ }
        searchBoxViewModel.onClickBackPressButton()

        // then
        assertEquals(1, finishEventCount)
    }

    @Test
    fun `뒤로가기 클릭시 포커스가 있다면 종료 이벤트가 호출되는지 않는지 테스트`() {
        // when
        var finishEventCount = 0
        searchBoxViewModel.searchBoxFinishEvent.observeForever { finishEventCount++ }
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickBackPressButton()

        // then
        assertEquals(0, finishEventCount)
    }

    @Test
    fun `뒤로가기 두번 클릭시 포커스가 존재했다면 종료 이벤트가 호출되는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(mutableListOf()))

        // when
        var finishEventCount = 0
        searchBoxViewModel.searchBoxFinishEvent.observeForever { finishEventCount++ }
        searchBoxViewModel.onClickSearchBox()
        searchBoxViewModel.onClickBackPressButton()
        searchBoxViewModel.onClickBackPressButton()

        // then
        assertEquals(1, finishEventCount)
    }

    @Test
    fun `키워드 삭제 버튼 클릭시 레파지토리에 삭제 요청을 하는지 테스트`() {
        // given
        every { mockImageSearchRepository.deleteSearchLog(any()) } returns (Completable.complete())

        // when
        val target = SearchLog("테스트", 1)
        searchBoxViewModel.onClickSearchLogDeleteButton(target)

        // then
        verify(exactly = 1) { mockImageSearchRepository.deleteSearchLog(target) }
    }

    @Test
    fun `키워드 삭제 버튼 클릭시 보유하던 검색 기록_목록에서 제거하는지 테스트`() {
        // given
        val searchLogList = createVirtualSearchLogList(3)
        val expectedList = createVirtualSearchLogList(3).apply { removeAt(0) }.sorted()

        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(searchLogList))
        every { mockImageSearchRepository.deleteSearchLog(any()) } returns(Completable.complete())

        // when
        searchBoxViewModel.loadSearchLogList()
        searchBoxViewModel.onClickSearchLogDeleteButton(searchLogList[0])

        // then
        searchBoxViewModel.searchLogList.observeForever{ receivedList ->
            assertEquals(expectedList, receivedList);
        }
    }

    @Test
    fun `키워드_검색_버튼_클릭시_기존에_검색했던_키워드라면_목록의_가장_앞쪽으로_이동시키는지_테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(createVirtualSearchLogList(3)))
        every { mockImageSearchRepository.insertOrUpdateSearchLog("테스트0") } returns (Single.just(SearchLog("테스트0", 4)))

        // when
        searchBoxViewModel.loadSearchLogList()
        searchBoxViewModel.onClickSearchButton("테스트0")

        // then
        searchBoxViewModel.searchLogList.observeForever{ searchLogs ->
            assertEquals("테스트0", searchLogs[0].keyword)
            assertEquals(3, searchLogs.size)
        }
    }

    @Test
    fun `실시간으로 키워드 변경 후 검색 버튼 클릭 시 검색 로그 저장 요청이 이루어지는지 테스트`() {
        // given
        every { mockImageSearchRepository.requestSearchLogList() } returns (Single.just(createVirtualSearchLogList(0)))
        every { mockImageSearchRepository.insertOrUpdateSearchLog("가나다") } returns (Single.just(SearchLog("가나다", 1)))

        // when
        searchBoxViewModel.onChangeKeyword("가")
        searchBoxViewModel.onChangeKeyword("가나")
        searchBoxViewModel.onChangeKeyword("가나다")
        searchBoxViewModel.onClickSearchButton()

        // then
        verify(exactly = 1) { mockImageSearchRepository.insertOrUpdateSearchLog("가나다") }
    }

    private fun createVirtualSearchLogList(size: Int): MutableList<SearchLog>  {
        val searchLogList = mutableListOf<SearchLog>()
        for(i in 0 until size) {
            searchLogList.add(SearchLog("테스트$i", i.toLong()))
        }

        return searchLogList
    }


}